README.md
# CssGaga Changelog

[CssGaga](http://www.99css.com/cssgaga/) 没有完整的更新日志，只是在更新的时候提示最近几次的更新日志，时间一长就不清楚有什么变动，于是我写了个脚本来获取完整的更新日志， 结果见 CHANGELOG.txt。

CssGaga 放在 [Google Code](http://cssgaga.googlecode.com) 上，[提交记录](https://code.google.com/p/cssgaga/source/list)。不过 [Google Code 快关闭了](http://google-opensource.blogspot.com/2015/03/farewell-to-google-code.html)。因为我也更新一下这个项目。之前是用 Python 写的，现在改用 Node.js 实现。

CssGaga 现在放在 [Github](https://github.com/ytzong/CssGaga) 上，[提交记录](https://github.com/ytzong/CssGaga/commits/gh-pages)。不过这里只抓取在 Google Code 上的更新记录。

## 使用说明

因为众所周知的原因，需要代理访问 Google。

事先安装好:

- Node.js, 这里用的是 io.js。
- Git 客户端，这里用的是 MsysGit。
- SVN 客户端, 这里用的是 TortoiseSVN。

打开 Git Bash，运行：

```bash
# 检查 svn 是否可用
svn --help
git clone https://github.com/yanxyz/cssgaga-changelog.git
cd cssgaga-changelog
svn checkout http://cssgaga.googlecode.com/svn/trunk/upload --depth empty cssgaga
npm i
node index.js -h
```

## 版权

MIT


output/changelog.txt
314

1. 修复某些 FTP 虚拟磁盘无法同步的 bug
2. 优化逻辑，文件处理之前先检查目标环境是否可用，以免徒劳一场



313

增加 SVN 更新 功能



312

修复 CSS 新文件名失效 bug（感谢 markqin 反馈）



311

修复 AutoRetina 对 gif 失效的 bug



310

新增关联 TortoiseSVN 功能（文件生成到目标路径后可自动更新 SVN 并弹出 SVN 提交窗口）

*注：
1. 请在更多选项中填写 TortoiseSVN 安装目录
2. 对应关系中在勾选 关联 TortoiseSVN 选项



309

皮肤 1toN 速度优化



308

修复第一次运行时缺少 dll 的 bug



307

1. 修复空白帧动画写法（例如0%{}）所导致的压缩出错 bug 
2. 修复文件打包时因文件重复不能打包的 bug



306

1. 界面输出处理时间
2. 临时压缩包更改后缀，防止被其他进程占用



305

修复某些情况下文件同步到临时目录的 bug



304

1. 速度优化
2. 去除 ini 功能



303

1. 调整 PNG32 转 PNG8(α) 后的文件名及位置（转换后的文件生成在当前目录下的 8a 目录中，且文件名与源文件名相同）
2. 修复图片合成功能中宽高多 1px 的 bug



302

1. 调整 PNG32 转 PNG8(α) 后的文件名及位置（转换后的文件生成在当前目录下的 8a 目录中，且文件名与源文件名相同）
2. 修复图片合成功能中宽高多 1px 的 bug



301

调整 AutoRetina 生成的 media query 格式，增加 min-resolution:1.25dppx (感谢 shenfeili 反馈)



300

增加 Icon 1 to N 功能



299

增加生成 iOS 及 Android 各种尺寸图片功能（内测，请联系 ytzong 开启）
修复启动时异常退出的 bug（感谢 nengzongtan 反馈）



298

回滚 PNG 压缩算法（感谢 tqtan 反馈新算法的问题）



297

使用 VS2013 重新编译



296

1. PNG 压缩算法改进
2. 修复 AutoSprite 中 .ie6.png 色块缺失 bug
3. 修复服务器在某些情况下启动时报错的 bug



295

1. PNG 压缩算法改进
2. 修复 AutoSprite 中 .ie6.png 色块缺失 bug
3. 修复服务器在某些情况下启动时报错的 bug



294

增加图片合成器



293

1. 增加 PSD TO PNG 功能
2. 提高皮肤 1toN 中的 AutoSrite 生成速度



292

1. 增加 PSD TO PNG 功能
2. 提高皮肤 1toN 中的 AutoSrite 生成速度



290

自动更新服务器迁移至 github



289

默认启动本地服务器，去除选项



288

增加时间戳格式设置



287

AutoRetina 格式改为
(-webkit-min-device-pixel-ratio: 1.25),
(min-resolution: 120dpi)



286

1. 更改新文件名规则，由每次自动命名改为手动刷新命名
2. 支持 css 新文件名，ini 格式调整
3. 增加文件类型设置，可自由配置需要同步的文件类型



285

base64 增加去除大小限制选项



284

修复当 css 多个选择器调用同一个 absolute 文件夹的图片时生成的 IE6 图片透明滤镜重复的 bug（感谢 yu xu 反馈）



283

修复注释没有正确关闭造成 CssGaga 死循环的 bug（感谢 webbwang 反馈）



282

渐变代码生成器支持 rgba



281

精简 RGBAtoFilter 的写法



280

修复配置文件写失败的 bug （其表现是某些功能虽然勾选了但不起作用）



279

修复 base64 bug



278

AutoRetina 功能支持 base64，规则是：
若本地有同名的 @2x 文件，则使用 @2x 的图片转码，并生成 background-size
也就是说不管屏幕是不是 retina，都调用 @2x 的 base64 图片



277

增加保存文件列表功能（更多选项中勾选保存文件列表开启）：
1. 源文件列表保存到 桌面\CssGaga\source-list-xxx.log
2. 提单文件列表保存到 桌面\CssGaga\report-list-n-xxx.log
注：其中 n 为目标路径序号，xxx 为日期
源文件列表拖到 CssGaga 中可以直接运行



276

增加代码生成器



275

界面调整



273

AutoRetina 格式调整（感谢 cheunglong 反馈）为
@media only screen and (-webkit-min-device-pixel-ratio:1.3),(min-resolution:96dpi) { ... }



272

AutoRetina 格式调整（感谢 cheunglong 反馈）为

@media only screen and (-webkit-min-device-pixel-ratio:1.5),(min-resolution:96dpi) { ... }



271

修复一些与算法无关的小 bug



270

支持 php 文件同步



269

界面调整
监听文件格式改为 css/htm/html/aspx



268

界面调整
监听文件格式改为 css/htm/html/aspx



267

界面调整
去除监听文件格式限制



266

去除 png32 有损压缩



265

调整 RGBAtoFilter 格式，感谢任平生的部分建议


源文件：.classname{ background-color:rgba(0,0,0,.5); }

Gaga 会添加一段兼容各个 IE 浏览器的代码：

.classname{background-color:transparent\0;filter:progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000);#zoom:1}
@media screen and (min-width:0\0){.classname{background-color:rgba(0,0,0,.5);-ms-filter:'none'}}



264

调整 RGBAtoFilter 格式，感谢任平生的建议


源文件：.classname{background-color:rgba(0,0,0,.5);}

Gaga 生成后会添加一段：

.classname {
	filter:progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000);
	#zoom:1;
}
:root .classname{
	-ms-filter: 'none';
}



263

调整 RGBAtoFilter 格式，感谢任平生及galendong的建议


源文件：.classname{background-color:rgba(0,0,0,.5);}

Gaga 生成后会添加一段：

.classname {
	filter:progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000);
	#zoom:1;
	#background-image:url(about:blank)
}
:root .classname{
	-ms-filter: 'none';
}



262

增加 RGBAtoFilter 功能（勾选开启）



261

增加 RGBAtoFilter 功能（勾选开启）

源文件：.classname{background-color:rgba(0,0,0,.5);}

Gaga生成后：

.classname{background-color:transparent\0;-ms-filter:'progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000)';#filter:progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000);#zoom:1}
@media screen and (min-width:0\0){.classname{background-color:rgba(0,0,0,.5);-ms-filter:'none';filter:none}}



260

增加 RGBAtoFilter 功能（勾选开启）

源文件：.classname{background-color:rgba(0,0,0,.5);}

Gaga生成后：

.classname{background-color:transparent\0;-ms-filter:'progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000)';#filter:progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000);#zoom:1}
@media screen and (min-width:0\0){.classname{background-color:rgba(0,0,0,.5);-ms-filter:none;filter:none}}



259

增加 RGBAtoFilter 功能（勾选开启）

源文件：.classname{background-color:rgba(0,0,0,.5);}

Gaga生成后：
.classname{background-color:transparent\0;-ms-filter:'progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000)';#filter:progid:DXImageTransform.Microsoft.gradient(startColorstr=#7F000000,endColorstr=#7F000000);#zoom:1}
@media screen and (min-width:0\0){.classname{background-color:rgba(0,0,0,.5);-ms-filter:none;filter:none}}



258

增加皮肤1to2功能，供调试时使用，皮肤1toN时同时勾选ASPX下面的调试即可



257

增加皮肤1to2功能，供调试时使用，皮肤1toN时同时勾选ASPX下面的调试即可



256

增加 PNG32 有损压缩功能，面积 <10000 或颜色数 <5000 的 PNG32 将会转换为 PNG8(α)
注：absolute 目录（自动生成 IE6 滤镜）的 PNG32 会跳过此功能



255

修复更新程序 bug



254

PNG 压缩参数调整



253

AutoSprite 新文件名规则调整为：
css名-当前时间小时数-每小时操作css的次数-当前机器用户名
例如 style.css，若选用新文件名（点更多在 beta 功能组中开启）
则生成的图片名为 style-16-0-ytzong.png



252

AutoSprite for retina 算法微调



251

1. 修复 bug：base64 空格 bug；import 格式 bug（感谢任平生反馈）
2. ini 格式调整（仅 Qzone 使用）
3. 屏蔽当路径含有 % 时文件不存在的警告（皮肤 1toN 时可以用 % 作为模板变量）
4. 更友好的配置提示（不需手动更改 <configuration> 节点，按提示重启程序即可；XP 用户未安装 .NET Framework 3.5 时自动打开下载链接；屏蔽配置正确者的配置提示）

test/parse.js
var checkout = require('../lib/checkout');
var parse = require('../lib/parse');
var should = require('should'); // jshint ignore:line

describe('parse.js', function() {
  this.timeout(5000);
  it('checkout r211', function(done) {
    checkout(211)
      .then(function() {
        return parse(211);
      })
      .then(function(result) {
        describe('parse 211.txt', function() {
          it('should have property', function() {
            result.should.have.property('241');
          });
        });
      })
      .then(done)
      .catch(function(err) {
        console.log(err.stack);
      });
  });
});


test/checkout.js
var checkout = require('../lib/checkout');
var should = require('should'); // jshint ignore:line

describe('checkout.js', function() {
  this.timeout(15000);
  it('should get four files', function(done) {
    checkout(116, 119).then(function(results) {
        results.length.should.be.equal(4);
        done();
      })
      .catch(function(err) {
        console.log(err);
      });
  });
});


index.js
#!/usr/bin/env node

// current.txt 记录了版本更新历史，因此任务就是得到并解析它。
// https://code.google.com/p/cssgaga/source/list?num=25&start=14
//
// SVN 只签出 current.txt
// 例如: svn update -r30 current.txt
//
// 将签出的 current.txt 放到 log 目录下面。
// 已有的就不再从服务器上签出。
//
// 有的 current.txt 里面的版本号不是连续的。
// 不同的修订版本有重叠的部分，而且还不一样。
// 按修订版本，新的覆盖旧的。
//
// current.txt 的编码有 gbk，也有 utf8（有 BOM 或没有）
// 统一保存为 utf8

'use strict';

var fsp = require('./lib/fsp');
var checkout = require('./lib/checkout');
var parse = require('./lib/parse');
var objectAssign = require('object-assign');

var argv = require('yargs')
  .options({
    t: {
      alias: 'test',
      type: 'boolean',
      describe: '测试合并本地的更新记录'
    },
    f: {
      alias: 'force',
      type: 'boolean',
      describe: '强制检查 google code 上的最后修订。'
    }
  })
  .help('h')
  .alias('h', 'help')
  .epilogue([
    'CssGaGa 已停止在 google code 上的更新，最后修订为 r314',
    '默认不检查最后修订，选项 -f 强制检查最后修订。'
  ].join('\n'))
  .argv;

try {
  fsp.mkdirSync('./output');
} catch (err) {

}

const CHANGLOG_FILE = './output/changelog.txt';
const JSON_FILE = './output/log.json';
const LAST_REV = 314;
var data;

if (argv.test) {
  fsp.deleteFile(CHANGLOG_FILE)
    .then(function() {
      return check(LAST_REV);
    });
} else if (argv.force) {
  check();
} else {
  check(LAST_REV);
}

function check(lastRev) {
  // 不存在 CHANGLOG_FILE 则重新检查
  return fsp.fileExists(CHANGLOG_FILE)
    .then(function(exists) {
      if (!exists) {
        return fsp.deleteFile(JSON_FILE);
      }
    })
    .then(function() {
      try {
        data = require(JSON_FILE);
      } catch (err) {
        data = {
          lastRev: 1, // 最后更新的修订版本。已知 r1 没有 current.txt, 跳过。
          revs: [] // 有 current.txt 的修订版本
        };
      }

      if (lastRev === data.lastRev) {
        console.log('no update');
        process.exit(0);
      }
    })
    .then(function() {
      return checkout(lastRev);
    })
    .then(function(rev) {
      if (rev === data.lastRev) {
        console.log('no update');
        process.exit(0);
      } else {
        let n = data.lastRev + 1;
        data.lastRev = rev;
        return checkout(n, rev);
      }
    })
    .then(function(revs) {
      return combine(revs);
    })
    .then(function() {
      return fsp.writeFileAsync(JSON_FILE, JSON.stringify(data));
    })
    .catch(function(err) {
      console.log(err.stack);
    });
}

function combine(revs) {
  return fsp.readFileAsync(CHANGLOG_FILE, 'utf8').then(parse)
    .catch(function() {
      // CHANGLOG_FILE 不存在，则重新解析已有的 txt
      revs.redo = true;
    })
    .then(function(obj) {
      if (revs.redo) {
        obj = {};
        revs = data.revs.concat(revs);
      }

      return Promise.all(revs.map(parse)).then(function(results) {
        results.forEach(function(x) {
          objectAssign(obj, x);
        });
        return obj;
      });
    })
    .then(function(obj) {
      data.revs = data.revs.concat(revs);
      var list = [];
      data.revs.forEach(function(x) {
        // 忽略没有内容的记录
        if (obj[x]) {
          list.push(x + rn(2) + obj[x]);
        }
      });
      // 倒序合并
      return fsp.writeFileAsync(CHANGLOG_FILE, list.reverse().join(rn(4)));
    });
}

function rn(n) {
  return new Array(n + 1).join('\r\n');
}


lib/fsp.js
'use strict';

var fs = require('fs');
var iconv = require('iconv-lite');
var jschardet = require('jschardet');
var Promise = require('bluebird'); // jshint ignore:line
Promise.promisifyAll(fs);

fs.createDir = function(dirname) {
  return fs.accessAsync(dirname).catch(function() {
    return fs.mkdirAsync(dirname);
  });
};

fs.deleteFile = function(filename) {
  return fs.unlinkAsync(filename)
    .catch(function() {
      // 忽略文件不存在的错误
    });
};

fs.convertToUtf8 = function(oldPath, newPath) {
  return fs.readFileAsync(oldPath).then(function(buf) {
    return iconv.decode(buf, jschardet.detect(buf).encoding);
  }).then(function(content) {
    return fs.writeFileAsync(newPath, content.trim());
  });
};

// 不能用 fs.access 去判断文件存在，因为被 svn 锁定。
fs.fileExists = function(filename) {
  return fs.statAsync(filename).then(function(stats) {
      return stats.isFile();
    })
    .catch(function() {
      return false;
    });
};

fs.P = Promise;

module.exports = fs;


lib/parse.js
'use strict';

var fsp = require('./fsp');
var path = require('path');

const LOG_DIR = path.join(__dirname, '../log');

module.exports = function(n) {
  if (typeof n === 'number') {
    return fsp.readFileAsync(path.join(LOG_DIR, n + '.txt'), 'utf8')
    .then(function(content) {
      return parse(content, n);
    });
  } else {
    return fsp.P.resolve(parse(n));
  }
};

function parse(str, n) {
  // 61.txt
  str = str.replace('117 重要更新！', '117\r\n重要更新！');

  var re = /^\s*\d+\s*$/mg;
  var vers = str.match(re);
  var contents = str.split(re);
  contents.shift();

  if (!vers || vers.length !== contents.length) {
    console.log(str, str.length, vers, contents);
    throw new Error(n + '.txt split wrong');
  }

  var ret = {};
  vers.forEach(function(x, i) {
    ret[x.trim()] = contents[i].trim();
  });
  return ret;
}


lib/checkout.js
'use strict';

var fsp = require('./fsp');
var path = require('path');
var Promise = require('bluebird'); // jshint ignore:line
var exec = Promise.promisify(require('child_process').exec);

const SVN_DIR = path.join(__dirname, '../cssgaga');
const LOG_DIR = path.join(__dirname, '../log');
const TXT_FILE = 'current.txt';
const TXT_PATH = path.join(SVN_DIR, TXT_FILE);

module.exports = function(fromRev, toRev) {
  if (!fromRev) {
    return checkoutLast();
  }

  if (!toRev || fromRev === toRev) {
    return fsp.createDir(LOG_DIR).then(function() {
      return checkout(fromRev);
    });
  }

  if (fromRev > toRev) {
    console.log(arguments);
    throw new Error('Wrong: fromRev > toRev');
  }

  let arr = [];

  for (let n = fromRev; n <= toRev; ++n) {
    arr.push(n);
  }

  // 串行，即一个个的签出，因为签出时 svn 锁定
  return fsp.createDir(LOG_DIR)
    .then(function() {
      return Promise.reduce(arr, function(arr, x) {
        return checkout(x).then(function(x) {
          if (x >= 0) {
            arr.push(x);
          }
          return arr;
        });
      }, []);
    });
};

// 签出指定的修订， 如果文件已存在则跳过。
// 返回有 current.txt 的修订版本号
function checkout(n) {
  var logFile = path.join(LOG_DIR, n + '.txt');
  return fsp.fileExists(logFile)
    .then(function(exists) {
      if (exists) {
        console.log('%d.txt already exists.', n);
        return n;
      }

      return exec('svn update -r' + n + ' ' + TXT_FILE, {
          cwd: SVN_DIR,
          timeout: 10000
        })
        .then(function() {
          return fsp.fileExists(TXT_PATH);
        })
        .then(function(exists) {
          if (exists) {
            return fsp.convertToUtf8(TXT_PATH, logFile)
              .then(function() {
                console.log('Updated to revision %d.', n);
                return n;
              });
          } else {
            console.log('Updated to revision %d, but not have %s', n, TXT_FILE);
          }
        });
    });
}

function checkoutLast() {
  return exec('svn update ' + TXT_FILE, {
      cwd: SVN_DIR,
      timeout: 5000
    })
    .then(function(output) {
      // console.log(output);
      var n = output[0].match(/\d+/)[0];
      // 必须是数字
      n = Number(n);
      var logFile = path.join(LOG_DIR, n + '.txt');
      return fsp.convertToUtf8(TXT_PATH, logFile)
        .then(function() {
          // console.log('Updated to revision %d.', n);
          return n;
        });
    });
}


get.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Ivan Yan 2013-12-20
# 获取并整合 CssGaGa 更新日志
#
# 2014-04-10
# 改为 python 3.4
#

from __future__ import unicode_literals

import os
import sys
import subprocess
import re
import codecs
import json
import argparse

changelog = 'CHANGELOG.txt' # 整合的日志
currenttxt = 'current.txt' # svn 日志
logdir = './log' # log 文件放置目录

if not os.path.exists(logdir):
	os.mkdir(logdir)

logjson = os.path.join(logdir, 'log.json')

def sav_config():
	open(logjson, 'w').write(json.dumps(log))

if not os.path.exists(logjson):
	log = {
		'r0': 0, # 远程的 revision
		'r': 0, # 已获取的 revision
		'n': 0, # 已处理的 revision
		'a': [] # 有 current.txt 的 revision
	}
	sav_config()
else:
	log = json.load(open(logjson, 'r'))


# 获取版本
# 版本在 current.txt 中， 只签出此文件

def check():
	'''检查 revision

	注意检查更新后没有自动获取更新
	'''
	output = subprocess.check_output(["svn", "up", currenttxt]).decode("latin")
	print(output)
	m = re.search(r'\d+', output)
	remote_r = int(m.group(0))
	tip = True

	if remote_r == log['r0']:
		if log['r'] == log['r0']:
			tip = False
			print('已经是最新的')
	else:
		log['r0'] = remote_r
		sav_config()
		print('有更新')

	if tip:
		print('使用参数 -u 来获取更新')

def up(num):
	'''获取指定的 revision

	更新如果包含 current.txt 会提示 Updated to revision xxx.
	有的更新没有，比如 r1 只是初始化目录，这时跳过
	https://code.google.com/p/cssgaga/source/list?num=25&start=6
	'''
	str_num = str(num)
	print('revision ' + str_num + " ..." )
	output = subprocess.check_output(["svn", "up", "-r" + str_num, currenttxt]).decode("latin")
	search = 'Updated to revision ' + str_num
	if search in output:
		logfile = os.path.join(logdir, str_num + '.txt')
		# 如果已存在则删除，保证可以重复运行
		if os.path.exists(logfile):
			os.remove(logfile)
		os.rename(currenttxt, logfile)
		if not num in log['a']:
			log['a'].append(num)
		print('done')
	else:
		print('skip')

def upall():
	'''获取全部 revision'''

	if log['r'] < log['r0']:
		print('开始获取日志...')
		for i in range(log['r'] + 1, log['r0'] + 1):
			up(i)
		log['r'] = log['r0']
		sav_config()
		print('已获取全部日志')
	else:
		if log['r0'] == 0:
			print('使用参数 -c 检查更新')
		else:
			print('已经是最新的')

# 按序从日志文件中提取版本
# 使用 map 结构： key 为版本号， value 为此版本号的更新信息
# 不同日志文件之间重复的版本，取后面的

def split_log(logfile, d = {}):
	# s = open(logfile, 'r').read()
	# 这里假定只有 gbk 与 utf-8 两种编码， 可能出错
	try:
		s = open(logfile, encoding='gbk').read()
	except UnicodeDecodeError:
		s = open(logfile, encoding='utf-8-sig').read()

	p = re.compile(r'^[ \t]*(\d+)[ \t]*\n', re.M)
	vers = p.findall(s)
	contents = p.split(s)
	for v in vers:
		j = contents.index(v)
		k = int(v)
		d[k] = contents[j+1].strip()
	return d

def save(d):
	content = ''
	keys = sorted(d)
	for k in keys:
		content = str(k) + '\n' + d[k] + '\n\n' + content
	open(changelog, mode='w', encoding='utf-8').write(content)

def merge_all():
	'''重新合并本地所有日志'''

	print('开始合并...')
	d = {}

	for i in range(len(log['a'])):
		logfile = os.path.join(logdir, str(log['a'][i]) + '.txt')
		d = split_log(logfile, d)

	save(d)
	print('合并完成')

def merge_up():
	'''增量合并'''

	if log['n'] < log['r']:
		print('开始合并...')
		d = split_log(changelog) if os.path.exists(changelog) else {}
		for i in range(log['n'] + 1, log['r'] +1):
			if i in log['a']:
				logfile = os.path.join(logdir, str(i) + '.txt')
				d = split_log(logfile, d)
		save(d)
		log['n'] = log['r']
		sav_config()
		print('合并完成')
	else:
		if log['r'] == 0:
			print('使用参数 -c 检查更新')
		else:
			print('没有新内容供合并')

def test():
	# content = open('log/119.txt', 'r').read()
	content = open('log/5.txt', 'r').read()
	# 处理 utf-8 bom
	# http://stackoverflow.com/questions/17912307/u-ufeff-in-python-string
	if content.startswith(codecs.BOM_UTF8):
		content = content.decode('utf-8-sig')

	p = re.compile(r'^[ \t]*(\d+)[ \t]*$', re.M)
	vers = p.findall(content)
	print(vers)
	contents = p.split(content)
	print(contents)

def test_save():
	log['a'] = [116, 117, 118, 119]
	merge_all()

if __name__ == '__main__':
	# test()

	parser = argparse.ArgumentParser(prog='getlog', description='获取并整合 CssGaga 更新日志')
	parser.add_argument('-c', action='store_true', help="检查更新")
	parser.add_argument('-u', action='store_true', help="获取检查到的更新")
	parser.add_argument('-a', action='store_true', help="重新合并本地所有更新日志")

	args = parser.parse_args()

	# 只支持一个选项
	if args.c:
		check()
	elif args.u:
		upall()
		merge_up()
	elif args.a:
		merge_all()
	else:
		parser.parse_args('-h'.split())
